cogs/poll.py

```diff
@@ -0,0 +1,318 @@
+        options_list = [[]]
+        page_number = 1
+        options_on_page = 0
+        char_on_page = len(question)
+        for arg in choices:
+            if options_on_page == 0 or (char_on_page + len(arg) < 550
+                                        and options_on_page < 7):
+                options_list[page_number - 1].append(arg)
+                options_on_page += 1
+                char_on_page += len(arg)
+            else:
+                page_number += 1
+                options_list.append([])
+                options_list[page_number - 1].append(arg)
+                options_on_page = 1
+                char_on_page = len(question) + len(arg)
+
+        embed.set_footer(
+            text=
+            "Initializing... Please wait, options will soon appear (Try 1/3)"
+            .format(len(options_list)))
+        msg = await ctx.send(embed=embed)
+
+        # add the emojis to the message
+        failed = False
+        tries = 0
+        while msg:
+            try:
+                pos = 0
+                for arg in choices:
+                    await msg.add_reaction(EMOJI_ALPHABET[pos])
+                    pos += 1
+                if len(options_list) > 1:
+                    await msg.add_reaction('◀')
+                    await msg.add_reaction('▶')
+                break
+            except discord.errors.Forbidden:
+                tries += 1
+                if tries == 3:
+                    failed = True
+                    break
+                else:
+                    embed.set_footer(
+                        text=
+                        "Was unable to put all reactions on message. Please don't add any emoji "
+                        "until this text disappears. (Try {}/3)".format(tries +
+                                                                        1))
+                    await msg.edit(embed=embed)
+                    await msg.clear_reactions()
+
+        # paginate
+        p = Pages(ctx, msg, options_list, embed, timeout, timeout_unit, failed)
+
+        await p.paginate()
+        return
+        # need to also eventually add options for anonymous voting in DMs
+        # restricting to only one choice only and
+        # add a duration for polls.
+
+
+class Pages:
+    def __init__(self,
+                 ctx,
+                 msg,
+                 options_list,
+                 embed,
+                 timeout,
+                 timeout_unit,
+                 failed=False):
+        self.bot = ctx.bot
+        self.user = ctx.author
+        self.message = msg
+        self.options_list = options_list
+        self.embed = embed
+        self.failed = failed
+        self.actions = [('◀', self._prev_page), ('▶', self._next_page)]
+        self.currentPage = 0
+        self.lastPage = len(options_list) - 1
+        self.timeout = timeout
+        self.timeout_unit = timeout_unit
+        self.timeout_end = time.time() + timeout * TIME_UNITS[timeout_unit]

```

Comment:
In almost all cases, the original paginator is used for displaying static content, so it's negligible if the bot fails while waiting to paginate. However, keeping longer timeouts means you need a better solution to deal with the bot shutting down randomly, and it's not really a good design to wait for, say, a day just to update a message. You don't have to handle all cases which rarely happen in reality. Is there a better way to handle timeout?

Reply:
The reason it takes a boolean "failed" as input is that since the poll puts a lot of emojis on the message, it's really easy to get to a point where users only need to put a couple emojis and the bot won't be able to paginate. And I don't think it's a rare thing because people on discord do like to put emojis on messages 😄 Before calling paginate, poll does try to take care of that by clearing emojis and trying again. However, if it isn't unable too, I still want the poll to print, which is why paginator gets called with failed=True. Paginator will then just print every options on the same page.

Reply:
That's not what I was talking about. I was referring to situations where the bot is shut down while a poll is active on the server

Reply:
Oh sorry, I see what you meant now. So I'd need to keep the message IDs of the polls in the database if I want to be able to make pages work again after the bot restarts?

Reply:
If yes this is is a good idea, because I will need to use the database anyway when I write anonymous polling by DM.

Reply:
Yes. Save the message id in the database and store it in memory on startup. I'm also suggesting that public voting should be done via a command so essentially the bot will not have to loop until poll timeout like it's doing now for paginating.

Reply:
Agreed about public voting via a command, but I still think that I should keep a simple version with emojis without paginator, maybe with a switch like "-s" for simple. Paginator anyway was already not that essential because people won't write polls that are very long, and even if they did the maximum is still less than 2000 characters, but I could even put a lower maximum for that mode

Reply:
Agreed about public voting via a command, but I still think that I should keep a simple version with emojis without paginator, maybe with a switch like "-s" for simple. Paginator anyway was already not that essential because people won't write polls that are very long, and even if they did the maximum is still less than 2000 characters, but I could even put a lower maximum for that mode

Reply:
I think it's nice to have something that just puts the emojis for you so you don't lose your vote but it's still simple and don't require people to use commands.

Reply:
Yeah. That would be great